### ISSUE TITLE
URLs are being truncated and then failing in Markdown, TOML files

### ISSUE THREAD

**Issue by rootwork** [urlstechie/urlchecker-action#94]
I can't understand how or why this might be happening, so I'll just describe what I'm seeing. I'm using `urlstechie/urlchecker-action@master`.

urlchecker-action is truncating _some_ URLs, trying to reach those truncated URLs, and reporting them as broken. Examples:

```md
[![linux](https://raw.githubusercontent.com/devicons/devicon/master/icons/linux/linux-original.svg)](https://www.linux.org/)
```

truncated to `https://www.linux`

```toml
  url = "https://codepen.io/rootwork/"
```

truncated to `https://codepen.io/root`

```md
[lazy-render things](https://www.drupal.org/node/1982024)
```

truncated to `https://www.drupal.org/node/19`

![truncated-urls](https://user-images.githubusercontent.com/79733/180121273-940fb0d0-3990-4ad8-9816-fa8bb22b0bc4.png)

urlchecker is looking at `.md`, `.toml`, `.yml` and `.scss` files. The ones that are being truncated are in the Markdown and TOML files but that might just be because there are many more of them.

These URLs are not split between lines; other than the first example above they are all actually pretty short URLs. The Markdown/TOML files validate. Many URLs, including other ones in those same files, are not truncated and pass urlchecker just fine.

Adjusting timeout and retry values on the linkchecker doesn't affect anything, because they're trying to check the wrong URLs.

All URLs were passing until July 12, at which point I introduced some broken URLs. After fixing them on July 13, I started getting the truncated URLs. (And to be clear the truncated ones aren't the ones that had been broken; they don't seem to have anything to do with each other.)

Any idea what might be happening?

**Comment by vsoch** [comment]
@rootwork I think I have an idea - can you give me (here in the comment) a quick markdown snippet with the failing urls and I'll work on fixing this asap?

**Comment by rootwork** [comment]
Sure. Here's the full paragraph containing the last example above, from a Markdown file. Linebreaks are in the original.

```markdown
FF: What I love the most about Twig is the syntax, and how it cleverly makes it
possible to [lazy-render things](https://www.drupal.org/node/1982024). The
possibilities of having an interpreted language are endless.
```

**Comment by vsoch** [comment]
Could you send an exact file from above? When I try to use that snippet it looks okay:

```bash
🤔 Uh oh... The following urls did not pass:
/home/vanessa/Desktop/Code/urlchecker-python/test.md:
     https://www.drupal.org/node/1982024
```
If I try the exact file and it doesn't reproduce it must be something about the GitHub environment, and I want to rule that out.

**Comment by vsoch** [comment]
And do you have a link to that run?

**Comment by rootwork** [comment]
Sure! [Here's the file](https://github.com/rootwork/rootwork.org/blob/main/hugo/content/blog/2013-05-16b/index.md), and [here's the run](https://github.com/rootwork/rootwork.org/runs/7425769180?check_suite_focus=true). But notice that it's inconsistent about which URLs get truncated; [here's a more recent run](https://github.com/rootwork/rootwork.org/actions/runs/2702377774) where there's still truncation, but that URL is not one of them.

**Comment by vsoch** [comment]
Ah! So I wonder if it's the colorizing of escape sequences? Check out the raw log here -  I don't think I see the full truncation if you scroll to the end? https://pipelines.actions.githubusercontent.com/serviceHosts/e1990149-fc7a-4131-8b50-bd5f555fd30c/_apis/pipelines/1/runs/618/signedlogcontent/2?urlExpires=2022-07-21T22%3A43%3A32.9120900Z&urlSigningMethod=HMACV1&urlSignature=24ReGiruxyomY876DUFpGBdCInSDm5AXvQXdGeVpkS4%3D

**Comment by vsoch** [comment]
okay so I can't share the URL - you can click or here is a screenshot comparison:

![image](https://user-images.githubusercontent.com/814322/180327401-3b214fb9-5253-4901-9be3-a5de074b2720.png)

![image](https://user-images.githubusercontent.com/814322/180327434-e1cab053-db69-4258-b191-3a4629e17e71.png)

**Comment by vsoch** [comment]
I'm going to do a quick test to look if ANSI escape sequences work better.

**Comment by vsoch** [comment]
okay I have a PR to urlchecker-python: https://github.com/urlstechie/urlchecker-python/pull/76
and you should be able to test the action via this branch `fix/ansi-colors` here https://github.com/urlstechie/urlchecker-action/pull/95

**Comment by rootwork** [comment]
Hmm. It seemed to have truncated fewer of them, but some are still there. [Here's the run.](https://github.com/rootwork/rootwork.org/runs/7459585626?check_suite_focus=true)

(Note one item in that list is a legitimate broken link, though.)

Update: [Run with the legitimately broken URL fixed.](https://github.com/rootwork/rootwork.org/runs/7459667404?check_suite_focus=true)

**Comment by vsoch** [comment]
Here is again the raw logs:

![image](https://user-images.githubusercontent.com/814322/180332179-bcd2f52e-98dc-467a-b062-eba8efd9fe79.png)

I can't say I've seen this before - I feel like it was working? I wonder if GitHub changed how they handle these sequences and our best shot is to remove all colors. The text is definitely there, it's just being hidden for some reason.

**Comment by vsoch** [comment]
and you are also saying there are false negatives? I've seen that too and I wonder if it's either our user agents aren't good enough (@SuperKogito was going to take a look at this) or if sites know to block requests from GitHub IPs.

**Comment by rootwork** [comment]
> and you are also saying there are false negatives?

No -- not apart from the truncated URLs. There are false negatives in the sense that the full URLs are valid and only fail after being truncated -- but it's not displaying any full URLs and saying they don't pass when they do.

**Comment by rootwork** [comment]
In my display of the run, I only see colored text on the files (`/github/workspace...` etc.) not on the links. Are the links supposed to be colored? If not, then maybe removing any color sequences around them?

Incidentally while I have it configured to do verbose reports (with the filenames) the same thing was happening even when it just listed the URLs.

**Comment by vsoch** [comment]
okay - so to debug let's first remove all coloring - give me a few mins to update the action container and I'll let you know when to re-run.

**Comment by vsoch** [comment]
okay updated container is pushed - try a new run! Should be digest:

```bash
sha256:cb4b2bff9af8b2beca9aea0ab54e5ecdd8c8b9556b6058fe6ff88ff34fd800c3 size: 1578
```
Let's see if just a print shows the entire url (and there is something wonky that needs further investigation around colors)

**Comment by rootwork** [comment]
Sorry, I'm not clear how to get at that -- should I still use `uses: urlstechie/urlchecker-action@fix/ansi-colors` in the GH action? Or `uses: urlstechie/urlchecker-action@cb4b2bff9af8b2beca9aea0ab54e5ecdd8c8b9556b6058fe6ff88ff34fd800c3`? This isn't my area of expertise 🤷🏽

**Comment by vsoch** [comment]
oh yes sorry - so you can use that same action branch, and the reason I provided the digest is just to check when it's pulling the action container that it grabs the right one! I think it should, but it's good to double check in case GitHub does something like cache the original one pulled.

**Comment by vsoch** [comment]
TLDR you should be able to just click the re-run :)

**Comment by rootwork** [comment]
[Didn't seem to change anything](https://github.com/rootwork/rootwork.org/runs/7460144545?check_suite_focus=true), unfortunately.

One thing I noticed, not sure it's significant -- the URLs that are listed are truncated, but if you hover over them they are actually linked to the correct (non-truncated) URLs. Not sure if that's just GH doing something outside of the action itself, though.

**Comment by vsoch** [comment]
This is so strange! I see the full URLs in the raw log:

![image](https://user-images.githubusercontent.com/814322/180336638-fb686f23-4b3e-4c94-afed-74d2f3e244c8.png)

I'm starting to wonder if GitHub is doing something strange with the URL and indent - I'm going to try and reproduce with a dummy Python script, and if I can I'll open an issue with GitHub. I'll keep you updated!

**Comment by rootwork** [comment]
OK, thanks! If there's anything else I can provide to help I'm happy to do so.

**Comment by vsoch** [comment]
Yep, looks like just from a python script with some prints, it tries to parse (and shorten) the url:

![image](https://user-images.githubusercontent.com/814322/180339502-04fb3809-8b81-471b-acd5-c949649a3b1d.png)

UNLESS it looks like you break some regex so it doesn't detect that it's a url?

![image](https://user-images.githubusercontent.com/814322/180339592-3bc0513a-0c5f-450b-86a5-c28dd4c40800.png)

So - I'm going to try adding a little arrow or something, and I'll report this to GitHub too.

**Comment by vsoch** [comment]
okay so I have a hypothesis - @rootwork give that workflow another re-run and let's see if we get a different result!

**Comment by rootwork** [comment]
Now it's reporting the full URLs, but still failing them. I'm guessing despite what it's reporting something is getting truncated in the background?

https://github.com/rootwork/rootwork.org/runs/7462969525?check_suite_focus=true

**Comment by vsoch** [comment]
Good - we fixed one bug!

I believe there are two issues here - printing the full url (fixed) and the response being 403 because the user agent or IP isn’t good enough to mimic a human (this issue we already knew about). There isn’t any truncation beyond the printing bug that now is fixed. And the latter is what @SuperKogito is working on, and let us know if you have ideas! In the meantime (tomorrow after work) I’ll put in a release to fix the first bug, and we will have to keep working on the second.

**Comment by rootwork** [comment]
OK, I think I understand what you're saying, but just to underscore that the URLs it is still wrongly failing are the same ones it was truncating before. I had thought it was failing them _because_ it was truncating them, but it sounds like you're saying there are other issues with it failing legitimate URLs, and it was only in printing those wrongly-failed URLs that truncation was happening.

If that's accurate then thank you for guiding me through this bug, and I'll look forward to fixes on the second. (I'll see if I can contribute anything.)

**Comment by vsoch** [comment]
Nope - it's not failing because of the truncation! It's failing because we need to update the headers, etc. that we use to make the request - servers are getting better at detecting our little test robot :) I have a bunch of ideas I will be testing likely over the weekend.

**Comment by vsoch** [comment]
And I agree it was a bit confounded and I totally see that we could make the connection from a truncation to the failure. But that's not the case because the URLs themselves were never actually truncated - that was just a printing bug in GitHub actions. In all cases we saw the full URLs in the raw logs.

**Comment by vsoch** [comment]
Did not mean to close this! So some of these URLs are fixed, the rest I'm going to try selenium soon! See https://github.com/urlstechie/urlchecker-python/issues/73

**Comment by vsoch** [comment]
Thanks again y'all - all URLs appear to be working now! Please open another issue if a new one arises.
